runtime/integration-tests/src/generic/envs/runtime_env.rs

```diff
@@ -0,0 +1,222 @@
+
+		Self {
+			nonce: 0,
+			ext: Rc::new(RefCell::new(ext)),
+			_config: PhantomData,
+		}
+	}
+
+	fn submit(&mut self, who: Keyring, call: impl Into<T::RuntimeCall>) -> DispatchResult {
+		self.ext.borrow_mut().execute_with(|| {
+			let runtime_call = call.into();
+			let info = runtime_call.get_dispatch_info();
+
+			let signed_extra = (
+				frame_system::CheckNonZeroSender::<T>::new(),
+				frame_system::CheckSpecVersion::<T>::new(),
+				frame_system::CheckTxVersion::<T>::new(),
+				frame_system::CheckGenesis::<T>::new(),
+				frame_system::CheckEra::<T>::from(Era::mortal(256, 0)),
+				frame_system::CheckNonce::<T>::from(self.nonce),
+				frame_system::CheckWeight::<T>::new(),
+				pallet_transaction_payment::ChargeTransactionPayment::<T>::from(0),
+			);
+
+			let raw_payload =
+				SignedPayload::new(runtime_call.clone(), signed_extra.clone()).unwrap();
+			let signature =
+				MultiSignature::Sr25519(raw_payload.using_encoded(|payload| who.sign(payload)));
+
+			let multi_address = (Address::Id(who.to_account_id()), signature, signed_extra);
+
+			let extrinsic =
+				<T::Block as Block>::Extrinsic::new(runtime_call, Some(multi_address)).unwrap();
+
+			self.nonce += 1;
+
+			T::apply_extrinsic(extrinsic).unwrap()
+		})
+	}
+
+	fn pass(&mut self, blocks: Blocks<T>) {
+		self.ext.borrow_mut().execute_with(|| {
+			let next = frame_system::Pallet::<T>::block_number() + 1;
+
+			let last_block = match blocks {
+				Blocks::ByNumber(n) => next + n,
+				Blocks::BySeconds(secs) => {
+					let blocks = secs / pallet_aura::Pallet::<T>::slot_duration();
+					if blocks % pallet_aura::Pallet::<T>::slot_duration() != 0 {
+						blocks as BlockNumber + 1
+					} else {
+						blocks as BlockNumber
+					}
+				}
+				Blocks::UntilEvent { limit, .. } => limit,
+			};
+
+			for i in next..last_block {
+				T::finalize_block();
+				Self::prepare_block(i);
+
+				if let Blocks::UntilEvent { event, .. } = blocks.clone() {
+					let event: T::RuntimeEventExt = event.into();
+					if frame_system::Pallet::<T>::events()
+						.into_iter()
+						.find(|record| record.event == event)
+						.is_some()
+					{
+						return;
+					}
+				}
+			}
+		})
+	}
+
+	fn state_mut<R>(&mut self, f: impl FnOnce() -> R) -> R {
+		self.ext.borrow_mut().execute_with(f)
+	}
+
+	fn state<R>(&self, f: impl FnOnce() -> R) -> R {
+		self.ext.borrow_mut().execute_with(|| {
+			transactional::with_transaction(|| {
+				// We revert all changes done by the closure to offer an inmutable state method
+				TransactionOutcome::Rollback::<Result<R, DispatchError>>(Ok(f()))
+			})
+			.unwrap()
+		})
+	}
+}
+
+impl<T: Runtime> RuntimeEnv<T> {
+	fn prepare_block(i: BlockNumber) {
+		let slot = Slot::from(i as u64);
+		let digest = Digest {
+			logs: vec![DigestItem::PreRuntime(AURA_ENGINE_ID, slot.encode())],
+		};
+
+		let header = Header {
+			number: i,
+			digest,
+			state_root: H256::default(),
+			extrinsics_root: H256::default(),
+			parent_hash: [69u8; 32].into(),
```

Comment:
I suppose the `69` is an arbitrary number. Maybe we want to take default instead to reduce unexpected behaviour likelihood?

Reply:
Yeah, I think it is an arbitrary number. It's the number used to configure the `frame_system` when genesis is the default. I need to check if I can actually put whatever there, or if not, I should link to the number in `frame_system`. 

Good point!

Reply:
Confirmed, it was a magic number. Nothing to do with the default from `frame_system`. Thanks for comment this!

Reply:
> it was a magic number


No pun intended?

Reply:
A magic pun then 😆